File: Dev/C++/Mascaret/src/VEHA/Kernel/Environment.cpp
#include "VEHA/Kernel/Environment.h"
#include "VEHA/Entity/Entity.h"
#include "VEHA/Kernel/InstanceSpecification.h"
#include "VEHA/Kernel/Model.h"
#include "HAVE/VirtualHuman.h"
#include <boost/foreach.hpp>

namespace VEHA
{

Environment::Environment(shared_ptr<Model> model) : _model(model)
{
}

Environment::~Environment()
{
}
void Environment::addInstance(shared_ptr<InstanceSpecification> value)
{
	_instances[value->getName()]=value;
}
void Environment::removeInstance(shared_ptr<InstanceSpecification> value)
{
	_instances.erase(value->getName());
}

void Environment::renameEntities(const vector<shared_ptr<Entity> >& entities,const vector<string>& newName)
{
	for(size_t i=entities.size();i--;)
	{
		//instance :
		_instances.erase(entities[i]->getName());
		entities[i]->rename(newName[i]);
		addInstance(entities[i]);
	}
	//la classe
	pair<string,shared_ptr<InstanceSpecification> >p;
	BOOST_FOREACH(p,_instances)
	{
		p.second->checkSlots();
	}

	//les values dans properties
	//
	//Mieux par event

	//model.register

}
vector<shared_ptr<Entity> > Environment::getEntities()
{
	vector<shared_ptr<Entity> > entities;
	map<string, shared_ptr<InstanceSpecification> >::iterator it=_instances.begin();
	while(it != _instances.end())
	{
		shared_ptr<Entity> ent=shared_dynamic_cast<Entity>(it->second);
		if(ent)
			entities.push_back(ent);
		it++;
	}
	return entities;
}
shared_ptr<Entity> Environment::getEntity(const string& name)
{
	return shared_dynamic_cast<Entity>(getInstance(name));
}
vector<shared_ptr<VirtualHuman> > Environment::getVirtualHumans()
{
	vector<shared_ptr<VirtualHuman> > entities;
	map<string, shared_ptr<InstanceSpecification> >::iterator it=_instances.begin();
	while(it != _instances.end())
	{
		shared_ptr<VirtualHuman> ent=shared_dynamic_cast<VirtualHuman>(it->second);
		if(ent)
			entities.push_back(ent);
		it++;
	}
	return entities;
}
shared_ptr<VirtualHuman> Environment::getVirtualHuman(const string& name)
{
	return shared_dynamic_cast<VirtualHuman>(getInstance(name));
}

string Environment::getUrl()
{
	return _url;
}
void Environment::setUrl(string url)
{
	_url=url;
}

}

File: Dev/CS/Mascaret/Mascaret/VEHA/Kernel/Environment.cs
using System;
using System.Collections.Generic;

namespace Mascaret
{
    public class Environment
    {

        private Zone mainZone;
        public Zone MainZone
        {
            get { return mainZone; }
            set { mainZone = value; }
        }


        private Model model = null;
        public Model Model
        {
            get { return model; }
            set { model = value; }
        }


        private Dictionary<String, InstanceSpecification> instances;
        public Dictionary<String, InstanceSpecification> InstanceSpecifications
        {
            get { return instances; }
            set { instances = value; }
        }

        public void addInstance(InstanceSpecification spec)
        {
            instances.Add(spec.name/*.ToLower()*/, spec);
        }

        public void removeInstance(InstanceSpecification spec)
        {
            instances.Remove(spec.name);
        }

        public InstanceSpecification getInstance(string name)
        {
            if (instances.ContainsKey(name/*.ToLower()*/))
                return instances[name/*.ToLower()*/];
            else
                return null;
        }


        protected string url;
        public string Url
        {
            get { return url; }
            set { url = value; }
        }

        public List<Entity> getEntities()
        {
            List<Entity> entities = new List<Entity>();
            foreach (InstanceSpecification curEntity in instances.Values)
            {
                if (curEntity is Entity)
                    entities.Add(curEntity as Entity);
            }
            return entities;
        }

        public Entity getEntity(string name)
        {
            if (instances.ContainsKey(name))
                return ((Entity)instances[name]);
            else
                return null;
        }

        public List<VirtualHuman> getVirtualHumans()
        {
            List<VirtualHuman> entities = new List<VirtualHuman>();
            foreach (InstanceSpecification curVirtualHuman in instances.Values)
            {
                if (curVirtualHuman is VirtualHuman)
                    entities.Add(curVirtualHuman as VirtualHuman);
            }
            return entities;
        }

        public VirtualHuman getVirtualHuman(string name)
        {
            if (instances.ContainsKey(name))
                return ((VirtualHuman)instances[name]);
            else
                return null;
        }

        public void renameEntities(List<Entity> entities, List<string> newName)
        {
            for (int i = 0; i < entities.Count; i++)
            {
                instances.Remove(entities[i].name);
                entities[i].rename(newName[i]);
                addInstance(entities[i]);
            }

            foreach (InstanceSpecification instance in instances.Values)
            {
                instance.checkSlots();
            }
        }


        public Environment(Model model)
        {
            instances = new Dictionary<string, InstanceSpecification>();
            this.model = model;
        }

        public void print()
        {
            //for (int i = 0; i < this.InstanceSpecifications.Keys.Count; i++)
            //	InstanceSpecifications.Values[i].print();
        }

    }
}

